lib/puppet/configurer.rb

```diff
@@ -235,9 +235,16 @@ def run(options = {})
   end
 
   def run_internal(options)
-    start = Time.now
     report = options[:report]
 
+    if options[:start_time]
+      startup_time = report.time - options[:start_time]
+      report.add_times(:startup_time, startup_time)
+
+      # Alter the report start time to include the agent startup time
+      report.time = options[:start_time]
```

Comment:
Could we set the correct start time when creating the report in https://github.com/puppetlabs/puppet/blob/0edd97e2a6cf5ed99d64676a60dfe509e52cbe78/lib/puppet/configurer.rb#L202? That way we wouldn't need to "post-process" the time. As far as I remember, we never pass in a report to the `Configurer#run` method except in tests. For example, the `agent`, `apply` and `device` applications don't.

Reply:
I have extended `Puppet::Transaction::Report.new` to receive an optional `start_time` parameter.